src/libraries/System.Security.Cryptography.Cose/ref/System.Security.Cryptography.Cose.cs

```diff
@@ -55,17 +55,17 @@ public abstract partial class CoseMessage
         public System.ReadOnlyMemory<byte>? Content { get { throw null; } }
         public System.Security.Cryptography.Cose.CoseHeaderMap ProtectedHeaders { get { throw null; } }
         public System.Security.Cryptography.Cose.CoseHeaderMap UnprotectedHeaders { get { throw null; } }
-        public static System.Security.Cryptography.Cose.CoseSign1Message DecodeSign1(byte[] cborPayload) { throw null; }
+        public static System.Security.Cryptography.Cose.CoseSign1Message DecodeSign1(ReadOnlySpan<byte> cborPayload) { throw null; }
     }
     public sealed partial class CoseSign1Message : System.Security.Cryptography.Cose.CoseMessage
     {
         internal CoseSign1Message() { }
         [System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")]
-        public static byte[] Sign(byte[] content, System.Security.Cryptography.AsymmetricAlgorithm key, System.Security.Cryptography.HashAlgorithmName hashAlgorithm, System.Security.Cryptography.Cose.CoseHeaderMap? protectedHeaders = null, System.Security.Cryptography.Cose.CoseHeaderMap? unprotectedHeaders = null, bool isDetached = false) { throw null; }
+        public static ReadOnlySpan<byte> Sign(ReadOnlySpan<byte> content, System.Security.Cryptography.AsymmetricAlgorithm key, System.Security.Cryptography.HashAlgorithmName hashAlgorithm, System.Security.Cryptography.Cose.CoseHeaderMap? protectedHeaders = null, System.Security.Cryptography.Cose.CoseHeaderMap? unprotectedHeaders = null, bool isDetached = false) { throw null; }
```

Comment:
Things returning data should still use `byte[]`.  Spans can be written to, but then they need to be provided as input.

We also probably don't want to remove the `byte[]` methods, but rather have them as ease of use overloads.

```suggestion
        public static byte[] Sign(ReadOnlySpan<byte> content, System.Security.Cryptography.AsymmetricAlgorithm key, System.Security.Cryptography.HashAlgorithmName hashAlgorithm, System.Security.Cryptography.Cose.CoseHeaderMap? protectedHeaders = null, System.Security.Cryptography.Cose.CoseHeaderMap? unprotectedHeaders = null, bool isDetached = false) { throw null; }
        public static byte[] Sign(byte[] content, System.Security.Cryptography.AsymmetricAlgorithm key, System.Security.Cryptography.HashAlgorithmName hashAlgorithm, System.Security.Cryptography.Cose.CoseHeaderMap? protectedHeaders = null, System.Security.Cryptography.Cose.CoseHeaderMap? unprotectedHeaders = null, bool isDetached = false) { throw null; }
```

A more rich spanification would additionally look like

```suggestion
        public static bool TrySign(ReadOnlySpan<byte> content, Span<byte> destination, System.Security.Cryptography.AsymmetricAlgorithm key, System.Security.Cryptography.HashAlgorithmName hashAlgorithm, out int bytesWritten, System.Security.Cryptography.Cose.CoseHeaderMap? protectedHeaders = null, System.Security.Cryptography.Cose.CoseHeaderMap? unprotectedHeaders = null, bool isDetached = false) { throw null; }
```

and possibly also

```suggestion
        public static int Sign(ReadOnlySpan<byte> content, Span<byte> destination, System.Security.Cryptography.AsymmetricAlgorithm key, System.Security.Cryptography.HashAlgorithmName hashAlgorithm, System.Security.Cryptography.Cose.CoseHeaderMap? protectedHeaders = null, System.Security.Cryptography.Cose.CoseHeaderMap? unprotectedHeaders = null, bool isDetached = false) { throw null; }
```

Reply:
> We also probably don't want to remove the byte[] methods, but rather have them as ease of use overloads.

Doesn't the implicit casts take care of that?

Reply:
> Doesn't the implicit casts take care of that?

Somewhat yes, somewhat no.

Firstly, it only takes care of it for the people who understand what spans are.  Currently, that's probably mostly limited to the people who contribute to the dotnet/runtime repository, so the `byte[]`-accepting overload is being friendlier to less experienced .NET developers.

Second, the implicit conversion turns `(byte[])null` into `ReadOnlySpan<byte>.Empty`.  Often we want the `byte[]`-accepting overload to be null-rejecting, to provide better a better debugging experience than seeing the empty span.

Once the input has been verified as not null (which could just be with the new `!!` parameter suffix) then casting to `ReadOnlySpan<byte>` and calling the actual implementation works.  For this case, since you need a `ReadOnlyMemory<byte>` for the reader, what I'd probably do is

```C#
public byte[] Sign(byte[] contents!!, ...) => Sign((ReadOnlyMemory<byte>)contents, ...);

public byte[] Sign(ReadOnlySpan<byte> contents, ...)
{
    pointer memory manager stuff
    {
        return Sign(manager.Memory, ...);
    }
}

private byte[] Sign(ReadOnlyMemory<byte> contents, ...)
{
   real code goes here.
}
```

That way `byte[]`-based callers don't have to go through pinning.

Reply:
Thanks for explaining. 

I have another question regarding `bool TrySign(ReadOnlySpan<byte> content, Span<byte> destination, ..., out int bytesWritten)`. 
Do we really want a try-pattern on that API? And when is supposed to return false, true, and when should it throw? 

Can we make it `int Sign(ReadOnlySpan<byte> content, Span<byte> destination, ...)` instead? There are [APIs that does this and don't provide a `bool Try*` signature](https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.getbytes?view=net-6.0).

Reply:
> Do we really want a try-pattern on that API? And when is supposed to return false, true, and when should it throw?

Span-writing Try methods all have the same shape:

* True means it succeeded.
* False means `destination.Length` was too small to hold the answer.
* Any other problem results in an exception.

> Can we make it int Sign(ReadOnlySpan<byte> content, Span<byte> destination, ...) instead? 

That was also in the list of possible Spanified API.  That one throws when the destination is too small, in addition to any reason that the Try version would have thrown.

Reply:
As we discussed offline, I've send changes to add a `TrySign` API; we discarded `int Sign(ReadOnlySpan<byte> content, Span<byte> destination, ...)` because there's currently no way for the user to compute the buffer size they will need.